aquadoggo/src/graphql/schema.rs

```diff
@@ -3,76 +3,112 @@
+        // Construct the document fields object which will be named `<schema_id>Field`.
+        let schema_field_name = fields_name(schema.id());
+        let mut document_schema_fields = Object::new(&schema_field_name);
+
+        // For every field in the schema we create a type with a resolver.
+        //
+        // TODO: We can optimize the field resolution methods later with a data loader.
+        for (name, field_type) in schema.fields() {
+            document_schema_fields =
+                build_document_field_schema(document_schema_fields, name.to_string(), field_type);
+        }
+
+        // Construct the document schema which has "fields" and "meta" fields.
+        let document_schema = build_document_schema(&schema);
+
+        // Register a schema and schema fields type for every schema.
+        schema_builder = schema_builder
+            .register(document_schema_fields)
+            .register(document_schema);
+
+        // Add a query object for each schema. It offers an interface to retrieve a single
+        // document of this schema by it's document id or view id. Its resolver parses and
+        // validates the passed parameters, then forwards them up to the children query fields.
+        root_query = build_document_query(root_query, &schema);
+
+        // Add a query for retrieving all documents of a certain schema.
+        root_query = build_all_document_query(root_query, &schema);
+    }
+
+    // Add next args to the query object.
+    let root_query = build_next_args_query(root_query);
+
+    // Build the schema.
+    schema_builder
+        .register(root_query)
         .data(store)
         .data(schema_provider)
         .data(tx)
         .finish()
-}
-
-/// Returns GraphQL API schema for p2panda node with a little trick to make dynamic schemas work.
-///
-/// The `async_graphql` crate we're using in this project does only provide methods to generate
-/// GraphQL schemas statically. Ideally we would like to query our database for currently known
-/// p2panda schemas and accordingly update the GraphQL schema whenever necessary but we don't have
-/// static and sync access to the database when building `async_graphql` types.
-///
-/// With this little workaround we are still able to make it work! We load the p2panda schemas from
-/// the database and write them into a temporary in-memory store. When `async_graphql` builds the
-/// GraphQL schema we can load from this store statically to build the schemas on the fly.
-async fn build_schema_with_workaround(shared: GraphQLSharedData) -> RootSchema {
-    // Store all application schemas from database into static in-memory storage
-    let all_schemas = shared.schema_provider.all().await;
-    save_static_schemas(&all_schemas);
-
-    // Build the actual GraphQL root schema, this will internally read the created JSON file and
-    // accordingly build the schema
-    build_root_schema(shared.store, shared.tx, shared.schema_provider)
+        .unwrap()
```

Comment:
What could go wrong here when unwrapping?

Reply:
Ah, it would be if we refer to a type which isn't registered I think, which we shouldn't be doing.